File: core/src/main/java/y2k/joyreactor/services/MessageService.java
package y2k.joyreactor.services;

import rx.Observable;
import y2k.joyreactor.Message;
import y2k.joyreactor.services.repository.MessageForUserQuery;
import y2k.joyreactor.services.repository.MessageThreadQuery;
import y2k.joyreactor.services.repository.Repository;
import y2k.joyreactor.services.synchronizers.PrivateMessageSynchronizer;

import java.util.List;

/**
 * Created by y2k on 12/8/15.
 */
public class MessageService {

    private PrivateMessageSynchronizer fetcher;
    private Repository<Message> repository;

    public MessageService(PrivateMessageSynchronizer fetcher, Repository<Message> repository) {
        this.fetcher = fetcher;
        this.repository = repository;
    }

    public Observable<List<Message>> getThreads() {
        return getFromRepo().mergeWith(
                fetcher.execute().flatMap(_void -> getFromRepo()));
    }

    private Observable<List<Message>> getFromRepo() {
        return repository.queryAsync(new MessageThreadQuery());
    }

    public Observable<List<Message>> getMessages(String username) {
        return repository.queryAsync(new MessageForUserQuery(username));
    }
}

File: core/src/main/kotlin/y2k/joyreactor/services/MessageService.kt
package y2k.joyreactor.services

import rx.Observable
import y2k.joyreactor.Message
import y2k.joyreactor.services.repository.DataContext
import y2k.joyreactor.services.synchronizers.PrivateMessageFetcher
import java.util.*

/**
 * Created by y2k on 12/8/15.
 */
class MessageService(
    private val fetcher: PrivateMessageFetcher,
    private val buffer: MemoryBuffer) {

    fun getThreads(): Observable<List<Message>> {
        return getFromRepo().mergeWith(fetcher.execute().flatMap { getFromRepo() })
    }

    private fun getFromRepo(): Observable<List<Message>> {
        val usersAlreadyAdded = HashSet<String>()
        val result = buffer.messages.filter { message ->
            if (usersAlreadyAdded.contains(message.userName))
                false
            else {
                usersAlreadyAdded.add(message.userName)
                true
            }
        }
        return Observable.just(result)
    }

    fun getMessages(username: String): Observable<List<Message>> {
        val result = buffer.messages.filter { it.userName == username }.sortedBy { it.date }
        return Observable.just(result)
    }
}